// Copyright 2021, Peter Birch, mailto:peter@lightlogic.co.uk
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/protobuf/empty.proto";

package NexusRPC;

service NXService {
    rpc Identify (google.protobuf.Empty) returns (NXIdentity) {}
    rpc Reset (google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc Parameters (google.protobuf.Empty) returns (NXParameters) {}
    rpc Status (google.protobuf.Empty) returns (NXStatus) {}
    rpc SetInterval (NXInterval) returns (google.protobuf.Empty) {}
    rpc SetActive (NXActive) returns (google.protobuf.Empty) {}
}

message NXIdentity {
    uint32 device_id     = 1;
    uint32 version_major = 2;
    uint32 version_minor = 3;
}

message NXParameters {
    uint32 counter_width  = 1;
    uint32 rows           = 2;
    uint32 columns        = 3;
    uint32 node_inputs    = 4;
    uint32 node_outputs   = 5;
    uint32 node_registers = 6;
}

message NXStatus {
    bool active        = 1;
    bool seen_idle_low = 2;
    bool first_tick    = 3;
    bool interval_set  = 4;
}

message NXInterval {
    uint32 interval = 1;
}

message NXActive {
    bool active = 1;
}
